scripts/test_path_preservation.ps1

```diff
@@ -0,0 +1,148 @@
+# This test installs emsdk and activates the latest toolchain using `--system` or `--permanent` flags, and checks if parts of PATH are lost or overwritten. Set $env:SYSTEM_FLAG and $env:PERMANENT_FLAG to test each. If no flag is provided the process/shell values are tested. See the CI file for an example.
```

Comment:
I think maybe there is still a bug here.

If I pass `--permanant` is still don't want any of my local PATH elements removed.

So I think that the `process/shell values` for PATH should be checked in all modes.

So I think when constucting the new PATH for the current shell we should always look in os.environ and not call `win_get_environment_variable` at all.

In other works we need to do two completely different PATH constructions.  Something like this:

```
environ_path = constuct_new_path_for_os_environ()
if permanant:
  registry_path = constuct_new_path_for_registry()
```

Notice that even in the registry mode we still need to construct a separate new path for the current environment.

Reply:
> So I think when constucting the new PATH for the current shell we should always look in os.environ and not call win_get_environment_variable at all.

With my current changes, we already do this. Also see https://github.com/emscripten-core/emsdk/pull/642#discussion_r515405800

Reply:
> If I pass --permanant is still don't want any of my local PATH elements removed.

The shell's PATH is not changed when `--permanent` is passed. If the user calls `refreshenv` then Windows updates the shell's PATH with the registry's PATH,

Reply:
I don't think that is accurate.   The calculated environment is always applied to the current shell.   On windows this is done by `write_set_env_script` which is called unconditionally. 

The idea is that emsdk is always activated in the current shell.  `--permanant` means that in addition we make the changes in the registry.

Reply:
You are trying to circumvent Windows behavior. Registry changes will apply to the shell, no matter what. This is baked in Windows. It may just be deferred depending on the shell that is used. There are some shells that automatically refresh the env changes.

Reply:
I'm merely describing the current (and desired) behaviour.  

"Registry changes will apply to the shell, no matter what." -- I'm not sure this is ture.  Registry changes apply to new shells.   The current shell will only be effected if you take specific action to refresh from the registry (and opt into clobbering any local change).

If I run a command prompt and call `set FOO=bar`, I do do not think its true that adding or changing `FOO` in the registry will effect my current shell.

I would suggest we we could simply call `refreshenv` on behalf of the user, but I do not think that is polite because they may have other local changes (e.g. local PATH additions) that they don't want to clobber. 

Please take a look at my proposed change and see what you think?  I am proposing that we construct two environments, one for the current shell and (optionally) one for the registry which is different.   I believe this is the behaviour that our users expect.

Reply:
> "Registry changes will apply to the shell, no matter what." -- I'm not sure this is ture. Registry changes apply to new shells. The current shell will only be effected if you take specific action to refresh from the registry (and opt into clobbering any local changes).

Well, there are some shells that watch the registry changes, and they call refreshenv.cmd in case of a change. This allows a long life shell. This technique is used in servers. 

> I would suggest we we could simply call refreshenv on behalf of the user, but I do not think that is polite because they may have other local changes (e.g. local PATH additions) that they don't want to clobber.

No. We should not. As I said, we should not circumvent Windows behavior.

> Please take a look at my proposed change and see what you think? I am proposing that we construct two environments, one for the current shell and (optionally) one for the registry which is different. I believe this is the behaviour that our users expect.

If you can merge this PR, my tests will show if those changes actually fix the problem. Once you merge this, you can add process test by pasting this code at [here](https://github.com/aminya/emsdk/blob/9b6acc9ee7cd917b196dd826c845010b34120fff/scripts/test_path_preservation.ps1#L81) and [here](https://github.com/aminya/emsdk/blob/9b6acc9ee7cd917b196dd826c845010b34120fff/scripts/test_path_preservation.ps1#L90)
```ps1
        if (Compare-Object -ReferenceObject $PATH_Process_BEFORE.Split(';') -DifferenceObject $PATH_Process.Split(';') ) {
            echo "Old process path is ...................."
            echo $PATH_Process_BEFORE
            echo "Current process path is ................"
            echo $PATH_Process
            throw "User PATH are changed while --system had been provided"
        }
```

I prefer a test-driven approach considering that this codebase is not small.